Bluetooth: Controller: Add NRF_ECB crypto support in nRF54L15 SoC

cvinayak opened the pull request:
Add NRF_ECB crypto support in nRF54L15 SoCs.

cvinayak committed:
Bluetooth: Controller: Add NRF_ECB crypto support in nRF54L15 SoC

Add NRF_ECB crypto support in nRF54L15 SoCs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Include mem and ecb in-system tests

Include mem and ecb implementation in-system tests.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nRF54L: Fix ecb tag name should be unique

Fix ecb tag name for violation to rule 5.7 (Tag name should
be unique) tag: ecb.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: bsim: Bluetooth: Add -RealEncryption=1 to pass in-system UT

BSIM tests that enabled CONFIG_BT_CTLR_TEST=y now require
-RealEncryption to pass the Controller's in-system unit test
of ECB module.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/ecb.c:
```diff
@@ -16,17 +17,82 @@
+
+		ep->in[0].ptr = ep->clear_text;
+		ep->in[0].length = sizeof(ep->clear_text);
+		ep->in[0].attribute = 7U;
```
Nit: Although this is already not a reserved attribute, the spec tells to use 11 (I think for no good reason). Anyhow, maybe this should be a define?

cvinayak replied:
Do you know if there is define already somewhere?

aescolar replied:
I don't

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/ecb.c:
```diff
@@ -108,39 +211,49 @@ uint32_t ecb_encrypt_nonblocking(struct ecb *ecb)
-	irq_disable(ECB_IRQn);
-}
+	if (NRF_ECB->EVENTS_ERRORECB) {
+#if defined(NRF54L_SERIES)
+		struct ecb *e = (void *)((uint8_t *)NRF_ECB->ECBDATAPTR -
+					 sizeof(struct ecb));
+#else /* !NRF54L_SERIES */
+		struct ecb *e = (void *)NRF_ECB->ECBDATAPTR;
+#endif /* !NRF54L_SERIES */
```
Suggestion: Maybe it would be a bit nicer with this , the task stop and irq_disable above the if?

cvinayak replied:
yes, agree that its nicer to move the refer closer to the callback being invoked.

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/ecb.c:
```diff
@@ -16,17 +17,82 @@
-		NRF_ECB->ECBDATAPTR = (uint32_t)ecb;
+
+#if defined(NRF54L_SERIES)
+		NRF_ECB->KEY.VALUE[3] = ((uint32_t)ep->key[0] << 24) |
+					((uint32_t)ep->key[1] << 16) |
+					((uint32_t)ep->key[2] << 8) |
+					(uint32_t)ep->key[3];
+		NRF_ECB->KEY.VALUE[2] = ((uint32_t)ep->key[4] << 24) |
+					((uint32_t)ep->key[5] << 16) |
+					((uint32_t)ep->key[6] << 8) |
+					(uint32_t)ep->key[7];
+		NRF_ECB->KEY.VALUE[1] = ((uint32_t)ep->key[8] << 24) |
+					((uint32_t)ep->key[9] << 16) |
+					((uint32_t)ep->key[10] << 8) |
+					(uint32_t)ep->key[11];
+		NRF_ECB->KEY.VALUE[0] = ((uint32_t)ep->key[12] << 24) |
+					((uint32_t)ep->key[13] << 16) |
+					((uint32_t)ep->key[14] << 8) |
+					(uint32_t)ep->key[15];
```
Suggestion: Maybe you want to put all this behind a macro or inline function and share it below? (in case you'd feel like optimizing this a bit later, so it would be easier to change this to be a plain copy and ifdef the reversing/not reversing in `ecb_encrypt[_be]`)

jhedberg commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/ecb.c:
```diff
@@ -16,17 +17,82 @@
-		NRF_ECB->ECBDATAPTR = (uint32_t)ecb;
+
+#if defined(NRF54L_SERIES)
+		NRF_ECB->KEY.VALUE[3] = ((uint32_t)ep->key[0] << 24) |
+					((uint32_t)ep->key[1] << 16) |
+					((uint32_t)ep->key[2] << 8) |
+					(uint32_t)ep->key[3];
+		NRF_ECB->KEY.VALUE[2] = ((uint32_t)ep->key[4] << 24) |
+					((uint32_t)ep->key[5] << 16) |
+					((uint32_t)ep->key[6] << 8) |
+					(uint32_t)ep->key[7];
+		NRF_ECB->KEY.VALUE[1] = ((uint32_t)ep->key[8] << 24) |
+					((uint32_t)ep->key[9] << 16) |
+					((uint32_t)ep->key[10] << 8) |
+					(uint32_t)ep->key[11];
+		NRF_ECB->KEY.VALUE[0] = ((uint32_t)ep->key[12] << 24) |
+					((uint32_t)ep->key[13] << 16) |
+					((uint32_t)ep->key[14] << 8) |
+					(uint32_t)ep->key[15];
```
Couldn't this be slightly simplified with:
```c
	NRF_ECB->KEY.VALUE[3] = sys_get_be32(&ep->key[0]);
	NRF_ECB->KEY.VALUE[2] = sys_get_be32(&ep->key[4]);
	NRF_ECB->KEY.VALUE[1] = sys_get_be32(&ep->key[8]);
	NRF_ECB->KEY.VALUE[0] = sys_get_be32(&ep->key[12]);
```

cvinayak replied:
Sure, this was a copy-paste from bare-metal implementations, will make the change...

jhedberg commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/ecb.c:
```diff
@@ -82,21 +148,58 @@ void ecb_encrypt(uint8_t const *const key_le, uint8_t const *const clear_text_le
 	/* setup the encryption h/w */
-	NRF_ECB->ECBDATAPTR = (uint32_t)ecb;
+#if defined(NRF54L_SERIES)
+	NRF_ECB->KEY.VALUE[3] = ((uint32_t)e->in_key_be[0] << 24) |
+				((uint32_t)e->in_key_be[1] << 16) |
+				((uint32_t)e->in_key_be[2] << 8) |
+				(uint32_t)e->in_key_be[3];
+	NRF_ECB->KEY.VALUE[2] = ((uint32_t)e->in_key_be[4] << 24) |
+				((uint32_t)e->in_key_be[5] << 16) |
+				((uint32_t)e->in_key_be[6] << 8) |
+				(uint32_t)e->in_key_be[7];
+	NRF_ECB->KEY.VALUE[1] = ((uint32_t)e->in_key_be[8] << 24) |
+				((uint32_t)e->in_key_be[9] << 16) |
+				((uint32_t)e->in_key_be[10] << 8) |
+				(uint32_t)e->in_key_be[11];
+	NRF_ECB->KEY.VALUE[0] = ((uint32_t)e->in_key_be[12] << 24) |
+				((uint32_t)e->in_key_be[13] << 16) |
+				((uint32_t)e->in_key_be[14] << 8) |
+				(uint32_t)e->in_key_be[15];
```
Same here

aescolar reviewed (dismissed):
Thanks. Just some minor suggestions/nits from me

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/ecb.c:
```diff
@@ -8,6 +8,8 @@
 #include <stdint.h>
 #include <string.h>
 
+#include <zephyr/sys/byteorder.h>
```
Shouldn't this commit be squashed on the first?

cvinayak committed:
Bluetooth: Controller: Refactor isr_ecb and unit testing

Refactor the NRF_ECB isr_ecb function to reduce redundant
common code.

Updated the unit tests to use Bluetooth Specification
defined test vectors.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>
